### DISCUSSION TITLE
Feature Request: Be able to download parts from Audible rather than whole books?

### DISCUSSION THREAD

**Discussion by digital-night** [mkb79/audible-cli#163] [Ideas]
Hi, I love this project, but when using it I get this message on some of my books:  

error: Error downloading [redacted].aax. Wrong  content type. Expected type(s): ['audio/aax', 'audio/vnd.audible.aax', 'audio/audible']; Got: text/html;charset=UTF-8; Message: This title is not supported for full audio file download, please download individual parts.

Looking online I can see a few people mention that they have run into this. Is it possible to add the ability to download the files as parts and combine on the machine once all parts of been fetched?  I've also had cases where the download gets interrupted before the entire book is downloaded, and I suspect if I could fetch and compile the parts that would make it more efficient to recover from failed downloads.

**Comment by mkb79** [comment]
@digital-night 
Thank you for your feature request. 

Yeah I can add the feature to allow downloading audiobooks in parts if necessary. You are willing to test this when it is ready?

Another feature on my todo list is to continue `paused` downloads. Do you have any knowledge in Python programming to realize this step?

**Comment by mkb79** [comment]
@digital-night 
If you want you can try [this](https://github.com/mkb79/audible-cli/tree/feat-resumeable-downloads) branch for testing the resume download function. 

*This will currently only work for aaxc files.*

How it works:
If an aaxc download aborts (e.g. CTRL+C, Internet Connection Reset) the temp file will be kept in the download directory. If you run the download command again, the temp file will be reused and `audible-cli` will download the missing content.

**Comment by mkb79** [comment]
@digital-night 

Do you have an asin from a book which have to downloaded in individual parts? I'll fix this but I need further information about content size and other things.

**Comment by Marzl7** [comment]
@mkb79 

Thank you for this great tool. I have the same problem with some of my books only available as part downloads. For example this one: 3837144054: Brandon Sanderson: Die Sturmlicht-Chroniken: Der Ruf der Klingen

If you could fix that problem it would be much appreciated. Happy new year!

**Comment by mkb79** [comment]
@Marzl7 
This and the resume download feature requests are next on my todo list. I'll take a look into the asin you are providing. Maybe I'll can give a first PR this week!

PS:
Ich wünsche ebenfalls noch ein frohes und gesundes Neues Jahr.

**Comment by mkb79** [comment]
@digital-night @Marzl7 
I've created [this](https://github.com/mkb79/audible-cli/tree/150-downloading-large-audiobooks) which should support downloading audiobooks in parts automatically. Maybe you can test these out?

**Reply by Marzl7** [reply]
Hey I run the command `pip install git+https://github.com/mkb79/audible-cli.git@150-downloading-large-audiobooks`  and than the download all command `audible download --all --aax-fallback`.

Still following error:  

> error: Error downloading Zeit_der_Finsternis_Blut_und_Knochen_3-LC_128_44100_stereo.aax. Wrong content type. Expected type(s): ['audio/aax', 'audio/vnd.audible.aax', 'audio/audible']; Got: text/html;charset=UTF-8; Message: This title is not supported for full audio file download, please download individual parts.

Did I miss a specific command for part downloading?

**Reply by mkb79** [reply]
Theoretically the download command should detect the error and download the audiobook in parts automatically. Let me try something.

**Reply by mkb79** [reply]
@Marzl7 
It seams I found the reason for the issue and fixed this. But there are one drawbacks. Downloading parts of aax files are not working. So it will now download aaxc files automatically. 

Please reinstall the branch with your command above and try again. 

Edit:
There was an issue in my last fix. Please reinstall from the branch again.

**Reply by Marzl7** [reply]
@mkb79 
Thx for the work.

Reinstalled and still an error if I run it with the `--aax-fallback` command. Wouldnt be a big problem do just download it with the axxc command but my library is so big, with the --all command it just skips the already existing files which is really convenient. Also as far as I remember the axxc cant be decrypted with ffmpeg, right?

Error: 
> error: _add_audioparts_to_queue() missing 1 required positional argument: 'aax_fallback'

**Reply by mkb79** [reply]
@Marzl7 
The error was fixed. Please reinstall with the same command and run it with `- -aax-fallback`. This should work.

**Reply by mkb79** [reply]
> Also as far as I remember the axxc cant be decrypted with ffmpeg, right?

You need at least ffmpeg 4.4 for decrypting aaxc files. [Here](https://github.com/mkb79/audible-cli/blob/master/plugin_cmds/cmd_decrypt.py) you can find a plugin script for audible-cli.

**Reply by Marzl7** [reply]
@mkb79 
> You need at least ffmpeg 4.4 for decrypting aaxc files. [Here](https://github.com/mkb79/audible-cli/blob/master/plugin_cmds/cmd_decrypt.py) you can find a plugin script for audible-cli.
Okay thats good.

Reinstalled an it added all the parts to the queue. But now following error occurs:
> error: B004UZXRH4 is not published. It will be available in 64259 days, 03h:11m:09s

**Reply by mkb79** [reply]
@Marzl7 

> error: B004UZXRH4 is not published. It will be available in 64259 days, 03h:11m:09s

Okay. It seams it detects the error correctly and it tries to download the parts. I'll have to take a look deeper into your audiobook. 

PS:
You can use e.g. `--start-date 2024-01-01` to download only books added to your library at this date or later. If you run `audible download --all --aaxc --start-date 2024-01-01` you only download books added at the specific date in aaxc format. So books before this date are ignored and not redownloaded.

**Reply by Marzl7** [reply]
@mkb79 
The Problem is that every book that should be downloaded in parts gets this error. I don't own books that are not released yet.

**Reply by mkb79** [reply]
All parts of your audiobook are all released at `2200-01-01T00:00:00Z`.  That are the `publication_datetime` values from the API response. The publication date is verified by audible-cli and will raise if not published. It's possible to override this behave. But this could lead in some circumstances for audiobooks which are really not published.

**Reply by mkb79** [reply]
@Marzl7 
You can now reinstall from branch again. Downloading in parts should now works. It did for me using your asin.

**Reply by Marzl7** [reply]
@mkb79 
Yes! Thank you its working great know.

**Reply by mkb79** [reply]
Great to hear that. Should I include an option to force the download in parts, if possible? Or is a single file (even if it is large) the better way?

PS:
Thank you for your support! I am very grateful for that.

**Reply by Marzl7** [reply]
I prefer the single file option when it's possible, makes it way easier to convert. But adding options is always good maybe someone needs the books in parts.
